### PR TITLE
Bugger off to better target locations

### PR DESCRIPTION
**Pull request by loveridge**
This branch is mainly looking to solve #905, with a few other changes attempting to get better bugger off target positions. 

Currently if the next command is a movement command, the target location will be towards that command position. Here we additionally check if the next movement command is a build command, and then move perpendicular so that the next building is not blocked. 

If there are no additional commands in the queue, then we attempt to move directly out of the circle as proposed in the issue. If that location is blocked, then rotations from the nearest axis are checked, as buildings are usually on a grid. Then if after a few buggeroff attempts we still haven't gotten out of the buggeroffradius, we fall back to just picking random locations as it works currently.

I tried a version that got the path cost from the pathmanager, but it was too slow to issue the number of path checks I wanted. I am curious if there are better ways to check if the path is blocked than how I am using `TestMoveSquare`, maybe issue a couple testmovesquares in a line to the target position?

### CODE REVIEW

**Comment by lhog** [comment]
Sounds like a good optimization. will review later in details.

File: rts/System/float3.h [hunk 1]
```diff
@@ -422,6 +422,25 @@ class float3
...
+	 * The result is only normalized if the input vector and self are normalized.
+	 * @return new vector with the result
+	 */
+	float3 rotate2D(const float3& rotationVector) const {
+		// https://blog.demofox.org/2014/12/27/using-imaginary-numbers-to-rotate-2d-vectors/
+		float nx = x * rotationVector.x - z * rotationVector.z;
+		float nz = x * rotationVector.z + z * rotationVector.x;
+		return float3{ nx, y, nz };
+	}
+
```
Comment (sprunk) [inline rts/System/float3.h:437]: why is `rotate2D` here in the header but `snapToAxis` in the other file? also perhaps this one could be explicitly `inline`?

Comment (loveridge) [reply rts/System/float3.h:437]: `snapToAxis` uses `Sign` which is from `SpringMath` and the header didn't import it. Indeed, probably a good candidate for inlining.

Comment (loveridge) [reply rts/System/float3.h:437]: The other float3 functions aren't inlined, should I just let the compiler decide like the others do?

Comment (sprunk) [reply rts/System/float3.h:437]: yea if nothing is already inlined then it's probably fine not to introduce that

File: rts/Sim/Units/CommandAI/MobileCAI.cpp [hunk 2]
```diff
@@ -1018,35 +1022,75 @@ void CMobileCAI::NonMoving()
...
 		return;
 
-	if (((owner->pos - buggerOffPos) * XZVector).SqLength() >= Square(buggerOffRadius * 1.5f))
+	// increase the target distance if continuing to fail to clear
+	float targetDistance = buggerOffRadius + owner->radius*(1 + 0.3*buggerOffAttempts);
+	if (((owner->pos - buggerOffPos) * XZVector).SqLength() >= Square(targetDistance))
 		return;
 
-	float3 buggerVec;
+	float3 buggerVec = ZeroVector;
```
Comment (lhog) [inline rts/Sim/Units/CommandAI/MobileCAI.cpp:1030]: This should be covered by the default constructor of `float3`

Comment (loveridge) [reply rts/Sim/Units/CommandAI/MobileCAI.cpp]: :+1: Removed the zerovector assignments.

File: rts/Sim/Units/CommandAI/MobileCAI.cpp [hunk 3]
```diff
@@ -1018,35 +1022,75 @@ void CMobileCAI::NonMoving()
...
+			if (!owner->moveDef->TestMoveSquare(nullptr, buggerPos, buggerVec) || !buggerPos.IsInMap()) {
+				buggerPos = -OnesVector;
+			}
+			if (commandQue[j].IsBuildCommand()) {
+				// try the other side
+				buggerPos = buggerOffPos + (-buggerDirection * (targetDistance));
+				if (!owner->moveDef->TestMoveSquare(nullptr, buggerPos, buggerVec) || !buggerPos.IsInMap()) {
+					buggerPos = -OnesVector;
+				}
+			}
```
Comment (lhog) [inline rts/Sim/Units/CommandAI/MobileCAI.cpp:1067]: Should this be reorganized for better readability?

```
if (commandQue[j].IsBuildCommand())
	buggerDirection *= -1.0f;

buggerPos = buggerOffPos + buggerDirection * targetDistance;

if (!owner->moveDef->TestMoveSquare(nullptr, buggerPos, buggerVec) || !buggerPos.IsInMap()) {
	buggerPos = -OnesVector;
}
```

**Comment by lhog** [comment]
Please remove extra parenthesis like in `(targetDistance)`

**Comment by lhog** [comment]
Overall it's a bit hard to track every detail of this PR, but I hope it's well tested and in case of bugs caused by this PR you can assume ownership to fix it.

**Comment by loveridge** [comment]
I am away on vacation this week. I'll address your comments next week. I accidentally hit the close button on my mobile device >.<

**Comment by loveridge** [comment]
Revisiting this MR I have simplified the logic by removing the piece that looks at future movement commands. I think it gives equally good results to just move to the closest position outside of the buggeroff radius. And the code becomes a bit simpler.

**Comment by sprunk** [comment]
The float3 additions can be generally useful so if they were a separate commit/PR i think they could get in with less effort.

**Comment by lhog** [comment]
Yeah, flag when you're done. Let's merge this one.

**Comment by loveridge** [comment]
> Yeah, flag when you're done. Let's merge this one.

It's ready

File: rts/Sim/Units/CommandAI/MobileCAI.cpp [hunk 4]
```diff
@@ -74,7 +74,6 @@ CR_REG_METADATA(CMobileCAI, (
 	CR_MEMBER(lastCommandFrame),
 	CR_MEMBER(lastCloseInTry),
 	CR_MEMBER(lastBuggerOffTime),
-	CR_MEMBER(numNonMovingCalls),
```
Comment (ashdnazg) [inline rts/Sim/Units/CommandAI/MobileCAI.cpp:77]: should be a rename and not a deletion, I think?

Comment (loveridge) [reply rts/Sim/Units/CommandAI/MobileCAI.cpp:77]: I agree

Comment (marcushutchings) [reply rts/Sim/Units/CommandAI/MobileCAI.cpp:77]: Thanks for spotting that. I've corrected this in the mainline.  https://github.com/beyond-all-reason/spring/commit/852e710368be303ae816d322a54281745c9c4c76

Comment (marcushutchings) [reply rts/Sim/Units/CommandAI/MobileCAI.cpp:77]: Sorry if that felt abrupt: we're going through a release cycle and I'm eager to see all issues pushed through and tested by the test team ASAP.

**Pull request merged**